src/Compilers/CSharp/Portable/Symbols/Source/SourceExtensionTypeSymbol.cs

```diff
@@ -395,5 +398,37 @@ internal static bool IsRestrictedExtensionUnderlyingType(TypeSymbol type)
 
             return false;
         }
+
+        private FieldSymbol? UnderlyingInstanceField
+        {
+            get
+            {
+                if (IsStatic)
+                {
+                    throw ExceptionUtilities.Unreachable();
+                }
+
+                var extendedType = GetExtendedTypeNoUseSiteDiagnostics(null);
+                if (extendedType is null)
+                {
+                    return null;
+                }
+
+                if (_lazyUnderlyingInstanceField is null)
+                {
+                    var field = new SynthesizedFieldSymbol(this, extendedType, WellKnownMemberNames.ExtensionFieldName);
```

Comment:
>new SynthesizedFieldSymbol(this, extendedType, WellKnownMemberNames.ExtensionFieldName); [](http://example.com/codeflow?start=32&length=88)

Should this field be explicitly aligned?

Reply:
Fields in structs use `LayoutKind.Sequential` by default. Extensions do the same:
```
        internal sealed override TypeLayout Layout
        {
            get
            {
                var data = GetDecodedWellKnownAttributeData();
                if (data != null && data.HasStructLayoutAttribute)
                {
                    return data.Layout;
                }

                if (this.TypeKind is TypeKind.Struct or TypeKind.Extension)
                {
                    // PROTOTYPE consider disallowing attribute for explicit layout on extension types
                    // CLI spec 22.37.16:
                    // "A ValueType shall have a non-zero size - either by defining at least one field, or by providing a non-zero ClassSize"
                    // 
                    // Dev11 compiler sets the value to 1 for structs with no instance fields and no size specified.
                    // It does not change the size value if it was explicitly specified to be 0, nor does it report an error.
                    return new TypeLayout(LayoutKind.Sequential, this.HasInstanceFields() ? 0 : 1, alignment: 0);
                }

                return default(TypeLayout);
            }
        }

```

Reply:
> Fields in structs use ```LayoutKind.Sequential``` by default.

Could you point me to a documentation confirming that the first instance field will start at offset 0 then?

Reply:
From ECMA CLI (II.10.1.2):

sequential: The CLI shall lay out the fields in sequential order, based on the order of the fields in the logical metadata table (§II.22.15).

[Rationale: [...] sequential layout is intended to instruct the CLI to match layout rules commonly followed by languages like C and C++ on an individual platform, where this is possible while still guaranteeing verifiable layout. [...]

From the C99 standard section 6.7.2.1 bullet point 13 (unfortunately, gated):

Within a structure object, the non-bit-field members and the units in which bit-fields reside have addresses that increase in the order in which they are declared. A pointer to a structure object, suitably converted, points to its initial member (or if that member is a bit-field, then to the unit in which it resides), and vice versa. There may be unnamed padding within a structure object, but not at its beginning.